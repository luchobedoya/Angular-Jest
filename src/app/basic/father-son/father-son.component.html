<h4>Hijo</h4>
<code>{{client | json}}</code>

<br><br>

<button 
    class="btn btn-primary"
    *ngIf="client"
    (click)="onChange(5)"
    data-test="btn-id"
    >
    Cambiar ID
</button>

<button
    *ngIf="client"
    class="btn btn-outline-danger"
    (click)="onDelete()"
    data-test="btn-delete"
    >
    Borrar Cliente
</button>