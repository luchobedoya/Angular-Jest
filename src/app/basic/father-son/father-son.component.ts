import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss']
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdate = new EventEmitter<Client>();

  onDelete(): void {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number): void {
    if(!this.client) return;
    //this.client.id = newId;
    this.client = { ...this.client, id }
    this.onClientUpdate.emit({...this.client});
  }
}
