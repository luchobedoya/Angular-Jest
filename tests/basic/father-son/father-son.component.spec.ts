import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe de hacer match con el snapshot', () => {
      expect(compiled).toMatchSnapshot();
  });

  test('No deben de aparecer botones si no hay cliente', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('Deben de aparecer 2 botones si hay cliente', () => {
    component.client = {id: 1, name: 'maria'};
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('Debe de emitir onDeleteClient con el boton de eliminar', () => {
    component.client = {id: 1, name: 'maria'};
    fixture.detectChanges();
    jest.spyOn(component.onDeleteClient, 'emit');
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    console.log('=>' ,btnDelete?.textContent);
    btnDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });


  test('Debe de emitir onClientUpdated con el boton de cambiar ID', () => {
    component.client = {id: 1, name: 'maria'};
    fixture.detectChanges();
    jest.spyOn(component.onClientUpdate, 'emit');
    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent(new Event('click'));
    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'maria'
    });
  });


  test('Debe de emitir onChangeClient con el ID especificado Si hay client', () => {
    jest.spyOn(component.onClientUpdate, 'emit');
    component.onChange(10);
    expect(component.onClientUpdate.emit).not.toHaveBeenCalled();
    component.client = {id: 1, name: 'maria'};
    fixture.detectChanges();
    component.onChange(10);
    expect(component.onClientUpdate.emit).toHaveBeenCalled();
    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'maria'
    })
  });


});
