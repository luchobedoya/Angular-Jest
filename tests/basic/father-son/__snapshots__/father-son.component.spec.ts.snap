// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h4>
    Hijo
  </h4>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
